hwilib/devices/jadepy/jade.py

```diff
@@ -0,0 +1,819 @@
+            else:
+                # Unknown/unhandled/unexpected message
+                logger.error("Unhandled message received")
+                device_logger.error(message)
+
+    def read_response(self, long_timeout=False):
+        while True:
+            try:
+                return self.read_cbor_message()
+            except EOFError as e:
+                if not long_timeout:
+                    raise
+
+    @staticmethod
+    def validate_reply(request, reply):
+        assert isinstance(reply, dict) and 'id' in reply
+        assert ('result' in reply) != ('error' in reply)
+        assert reply['id'] == request['id'] or \
+            reply['id'] == '00' and 'error' in reply
+
+    def make_rpc_call(self, request, long_timeout=False):
+        # Write outgoing request message
+        assert isinstance(request, dict)
+        assert 'id' in request and len(request['id']) > 0
+        assert 'method' in request and len(request['method']) > 0
+        assert len(request['id']) < 16 and len(request['method']) < 32
+        self.write_request(request)
+
+        # Read and validate incoming message
+        reply = self.read_response(long_timeout)
+        self.validate_reply(request, reply)
+
+        return reply
+
+
+#
+# Low-level Serial backend interface to Jade
+# Calls to send and receive bytes over the interface.
+# Intended for use via JadeInterface wrapper.
+#
+# Either:
+#  a) use via JadeInterface.create_serial() (see JadeInterface)
+# (recommended)
+# or:
+#  b) use JadeSerialImpl() directly, and call connect() before
+#     using, and disconnect() when finished,
+# (caveat cranium)
+#
+class JadeSerialImpl:
+    def __init__(self, device, baud, timeout):
+        self.device = device
+        self.baud = baud
+        self.timeout = timeout
+        self.ser = None
+
+    def connect(self):
+        assert self.ser is None
+
+        logger.info('Connecting to {} at {}'.format(self.device, self.baud))
+        self.ser = serial.Serial(self.device, self.baud,
+                                 timeout=self.timeout,
+                                 write_timeout=self.timeout)
+        assert self.ser is not None
+        self.ser.__enter__()
+        logger.info('Connected')
+
+    def disconnect(self):
+        assert self.ser is not None
+        self.ser.__exit__()
+
+        # Reset state
+        self.ser = None
+
+    def write(self, bytes_):
+        assert self.ser is not None
+        return self.ser.write(bytes_)
+
+    def read(self, n):
+        assert self.ser is not None
+        return self.ser.read(n)
+
+
+#
+# Low-level BLE backend interface to Jade
+# Calls to send and receive bytes over the interface.
+# Intended for use via JadeInterface wrapper.
+#
+# Either:
+#  a) use via JadeInterface.create_ble() (see JadeInterface)
+# (recommended)
+# or:
+#  b) use JadeBleImpl() directly, and call connect() before
+#     using, and disconnect() when finished,
+# (caveat cranium)
+#
+class JadeBleImpl:
```

Comment:
I'm guessing we wouldn't be using this?

Reply:
Right - at the top of the file I commented out the dependencies it needs to try to reduce transitive dependency creep (and didn't include them in setup/project files).  
We could remove all the code entirely - but it'll make refreshing the lib from jade's tree that much more involved.

Reply:
You could split the source code into different python modules - `jade_bluetooth.py`, `jade_serial.py`, then import these two into `jade.py`.

Then the matter of refreshing the lib would be reduced to not copying `jade_bluetooth.py` and removing the `from jade_bluetooth import JadeBleImpl` from `jade.py`.

Reply:
Updated Jade repo as suggested, and copied here.